### DISCUSSION TITLE
Can custom nodes mark themselves as non-deletable (or non-draggable, non-selectable, etc)?

### DISCUSSION THREAD

**Discussion by jreyesr** [xyflow/xyflow#3036] [Q&A] [answered]
Hello! I'm currently working with custom nodes, and I've found something that I don't know how to do.

Can a custom node class declare itself as non-deletable? I'm aware of the [deletable](https://reactflow.dev/docs/api/nodes/node-options/#deletable) property, but that has to be declared in the node properties, and can't come from the node type.

My usecase is an [activity diagram](https://www.visual-paradigm.com/guide/uml-unified-modeling-language/what-is-activity-diagram/) editor, which should always have a single Start node (the solid black circle). I have a [StartNode custom node](https://reactflow.dev/docs/guides/custom-nodes/), but it's possible to delete it. I can certainly ensure that the node's declaration in the `initialNodes` array has the `deletable` prop set to `false`:

```js
const initialNodes: Node[] = [
  {
    id: "0",
    type: "start",
    position: { x: 0, y: 0 },
    data: {},
    deletable: false
  },
];
```

but it would be cleaner if that could be configured in the function that declares the custom node class, so that _every_ node that has the `start` type already comes with the `deletable` prop set to `false`.

Also, that is valid for all other similar props: `draggable`, `selectable`, `connectable` and `deletable`, along with `hidden`. Can any of those properties be controlled from "inside" the custom node's declaration? Or is it necessary to control them from "outside", from the `initialNodes` array?

Many thanks!

**Comment by moklick** [comment, marked as answer]
There are many ways to achieve this but there is no built-in functionality. I would create a helper function like `createNode(nodeType)` that returns a node object with the corresponding flags based on the node type. This function could also be used for your initial nodes:

```js 
const initialNodes: Node[] = [createNode('start')];
```

**Reply by jreyesr** [reply]
Yeah, that seems like the best way.

For reference, I ended up solving it by attaching a "builder function" to the actual custom node, as [JS lets you add properties on Function objects](https://stackoverflow.com/q/48475902):

```tsx
// file nodes/StartNode.tsx

function StartNode(props) {
  return (
    <>
      {/* Whatever the node contains */}
    </>
  );
}

StartNode.TypeKey = "start";

StartNode.Builder = function(): Node {
  // This is a builder function that creates 
  return {
    id: "start",
    position: {x: 0, y: 0},
    data: {},
    type: StartNode.TypeKey,
    deletable: false,
  };
};

export default StartNode;
```

`StartNode` is my custom node, it's a normal function. However, you can also get `StartNode.TypeKey` and `StartNode.Builder()`. Thus, on the main editor instance, you can run:

```tsx
const nodeTypes = {
  [StartNode.TypeKey]: StartNode,
};
const initialNodes: Node[] = [
  StartNode.Builder(), // This calls the builder func that returns a prebuilt node config
];

<ReactFlow
  ...
  nodeTypes={nodeTypes}
/>
```

That's essentially the same as recommended above, but now the `createNode` function is called `Builder` and is forever attached to the node type.

Thank you, @moklick!
